pennylane/shadows/classical_shadow.py

```diff
@@ -0,0 +1,140 @@
+
+        U = np.empty((T, n, 2, 2), dtype="complex")
+        for i, u in enumerate(self.observables):
+            U[np.where(recipes == i)] = u
+
+        state = (
+            qml.math.cast((1 - 2 * bitstrings[:, :, None, None]), np.complex64) * U + np.eye(2)
+        ) / 2
+
+        return 3 * state - np.eye(2)[None, None, :, :]
+
+    @staticmethod
+    def _obtain_global_snapshots(local_snapshot):
+        global_snapshots = []
+        for T_snapshot in local_snapshot:
+            tensor_product = T_snapshot[0]
+            for n_snapshot in T_snapshot[1:]:
+                tensor_product = np.kron(tensor_product, n_snapshot)
+            global_snapshots.append(tensor_product)
+
+        return np.array(global_snapshots)
+
+    def global_snapshots(self, wires=None, snapshots=None):
+        """Compute the T x 2**n x 2**n global snapshots"""
+
+        local_snapshot = self.local_snapshots(wires, snapshots)
+
+        if local_snapshot.shape[1] > 16:
+            warnings.warn(
+                "Querying density matrices for n_wires > 16 is not recommended, operation will take a long time",
+                UserWarning,
+            )
+
+        return ClassicalShadow._obtain_global_snapshots(local_snapshot)
+
+    def _convert_to_pauli_words(self, observable):
+        """Given an observable, obtain a list of coefficients and Pauli words, the
+        sum of which is equal to the observable"""
+
+        num_wires = self.bitstrings.shape[1]
+        obs_to_recipe_map = {"PauliX": 0, "PauliY": 1, "PauliZ": 2}
+
+        def pauli_list_to_word(obs):
+            word = [-1] * num_wires
+            for ob in obs:
+                if ob.name not in obs_to_recipe_map:
+                    raise ValueError("Observable must be a linear combination of Pauli observables")
+
+                word[ob.wires[0]] = obs_to_recipe_map[ob.name]
+
+            return word
+
+        if isinstance(observable, (qml.PauliX, qml.PauliY, qml.PauliZ)):
+            word = pauli_list_to_word([observable])
+            return [(1, word)]
+
+        if isinstance(observable, qml.operation.Tensor):
+            word = pauli_list_to_word(observable.obs)
+            return [(1, word)]
+
+        # TODO: cases for new operator arithmetic
+
+        if isinstance(observable, qml.Hamiltonian):
+            coeffs_and_words = []
+            for coeff, op in zip(observable.data, observable.ops):
+                coeffs_and_words.extend(
+                    [(coeff * c, w) for c, w in self._convert_to_pauli_words(op)]
+                )
+            return coeffs_and_words
+
+    def expval(self, H, k):
+        """
+        Get the expectation of an observable
+        """
+        coeffs_and_words = self._convert_to_pauli_words(H)
+
+        expval = 0
+        for coeff, word in coeffs_and_words:
+            expvals = pauli_expval(self.bitstrings, self.recipes, np.array(word))
+            expval += coeff * median_of_means(expvals, k)
+
+        return expval
```

Comment:
In the current version, the `expval` is returning complex valued integers, whereas qnodes return real values tensors. Would it make sense to cast the output at some point?

Reply:
I'm not sure where you're seeing complex-valued integers. Calling `shadow.expval(...)` returns a `float64` real number.

Reply:
Sorry that was a typo, I just meant complex numbers:
```
>>> shadow.expval(qml.PauliZ(0) @ qml.PauliZ(1), k=1)
(1.224+0j)
```

Reply:
Can you attach your full code here? I can't reproduce the problem locally:
```python
def qft_circuit(wires, shots=10000, interface="autograd"):
    dev = qml.device("default.qubit", wires=wires, shots=shots)

    one_state = np.zeros(wires)
    one_state[-1] = 1

    @qml.qnode(dev, interface=interface)
    def circuit():
        qml.BasisState(one_state, wires=range(wires))
        qml.QFT(wires=range(wires))
        return qml.classical_shadow(wires=range(wires))

    return circuit

def print_expval():
    circuit = qft_circuit(4)
    bits, recipes = circuit()

    shadow = ClassicalShadow(bits, recipes)
    print(shadow.expval(qml.PauliZ(0) @ qml.PauliZ(1), k=1))
```
```pycon
>>> print_expval()
0.0162
```
The expval computation doesn't even use complex numbers. It just find the recipes that match the Pauli word and averages across the bits.

Reply:
It is indeed strange that I get complex outputs with this example code 🤔 
```python
dev = qml.device("default.qubit", wires=range(2), shots=1000)
@qml.qnode(dev)
def qnode(x):
    qml.Hadamard(0)
    qml.CNOT((0,1))
    qml.RX(x, wires=0)
    return classical_shadow(wires=range(2))

bitstrings, recipes = qnode(0)

qnode_exact(0)
shadow = ClassicalShadow(bitstrings, recipes)
>>> shadow.expval(qml.PauliZ(0) @ qml.PauliZ(1), k=1)
(1.008+0j)
```

Reply:
I ran your code exactly as it is and got a real output:
```
>>> shadow.expval(qml.PauliZ(0) @ qml.PauliZ(1), k=1)
0.936
```

Reply:
🤯 
And you are also on the `shadow_class` branch? Thats so strange 🤔

Reply:
I added new tests to check if the dtype is `float64`, and they are passing.